code/datums/json_savefile.dm

```diff
@@ -0,0 +1,145 @@
+#define JSON_SAVEFILE_PARSE_CUTOFF_FLAG "||__||__||"
+#define JSON_SAVEFILE_PARSE_REPLACE_COMMA "|||_|||"
+
+/**
+ * A savefile implementation that handles all data using json
+ * also saves it using JSON too, fancy
+ */
+/datum/json_savefile
+	var/path
+	VAR_PRIVATE/list/tree
+	var/auto_save = FALSE
+
+/datum/json_savefile/New(path)
+	src.path = path
+	tree = list()
+	if(fexists(path))
+		load()
+
+/**
+ * Gets an entry from the json tree, with an optional default value.
+ * If no key is specified it throws the entire tree at you instead
+ */
+/datum/json_savefile/proc/get_entry(key, default_value)
+	if(!key)
+		return tree
+	return (key in tree) ? tree[key] : default_value
+
+/// Sets an entry in the tree to the given value
+/datum/json_savefile/proc/set_entry(key, value)
+	tree[key] = value
+	if(auto_save)
+		save()
+
+/// Removes the given key from the tree, if no key is given it clears the entire tree
+/datum/json_savefile/proc/clear(key)
+	if(key)
+		tree -= key
+	else
+		tree.Cut()
+	if(auto_save)
+		save()
+
+/datum/json_savefile/proc/load()
+	if(!fexists(path))
+		return FALSE
+	try
+		tree = json_decode(file2text(path))
+		return TRUE
+	catch(var/exception/err)
+		stack_trace("failed to load json savefile at '[path]': [err]")
+		return FALSE
+
+/datum/json_savefile/proc/save()
+	if(fexists(path))
+		fdel(path)
+	rustg_file_write(json_encode(tree), path)
```

Comment:
this means that now character preferences have a hard dependency on rust-g that did not exist before by the way.

Reply:
we're using this pattern everywhere tbf. fdel then write is a stupid pattern that is necessary in pure byond

Reply:
but they forgot to remove the fdel

Reply:
I am not aware of anywhere where it's a hard dependency, there are usually fallbacks to standard write where rust-g isn't available

Reply:
> this means that now character preferences have a hard dependency on rust-g that did not exist before by the way.

i always envisioned an abstracted system for writing preferences to disk that would let us plug in json or save files based on what child was instenced.

with datum preferences there really isn't any excuse to not do so so this can just fall back to savefiles. everything is already abstracted.

Reply:
fwiw theres more to savefiles than preferences and the stuff thats there is not datums yet